src/plugins/explore/public/application/utils/state_management/actions/detect_optimal_tab/detect_optimal_tab.ts

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { createAsyncThunk } from '@reduxjs/toolkit';
+import { RootState } from '../../store';
+import { setActiveTab } from '../../slices';
+import { ExploreServices } from '../../../../../types';
+import { defaultPrepareQuery } from '../query_actions';
+import { getVisualizationType } from '../../../../../components/visualizations/utils/use_visualization_types';
+
+/**
+ * Determine if results can be visualized
+ */
+const canResultsBeVisualized = (results: any): boolean => {
+  if (!results?.hits?.hits || !results?.fieldSchema || results.hits.hits.length === 0) {
+    return false;
+  }
+
+  const rows = results.hits.hits;
+  const fieldSchema = results.fieldSchema;
+  const visualizationData = getVisualizationType(rows, fieldSchema);
+
+  return !!visualizationData?.visualizationType;
+};
+
+/**
+ * Determine the optimal tab based on results
+ * Use visualization tab if we can find a visualization type
+ * Otherwise, fallback to logs tab
+ */
+const determineOptimalTab = (results: any): string => {
+  if (canResultsBeVisualized(results)) {
+    return 'explore_visualization_tab';
```

Comment:
is this always what user wants? for example, I might be in the logs tab because I want to compare numbers, and having to manually switch tab is not convenient

Reply:
well this is the current decided behavior: when switch tab or load saved explore, will respect customer's choice. When user run a new query, initialize explore or switch dataset, will auto detect. I asked the same question but it seems we will follow this pattern and ask user to feedback on this behavior.